jobs/audit_jobs.py

```diff
@@ -129,3 +176,96 @@ def expand(self, models_of_kind):
+        self._kind = kind
+
+    def expand(self, models_of_kind):
+        """Returns a PCollection of (key, count) pairs for each input model.
+
+        Args:
+            models_of_kind: PCollection. The input models.
+
+        Returns:
+            PCollection. A (key, count) pair for each input model. The key is a
+            tuple of (kind: str, ID: str), and the count is always be 1.
+        """
+        return (
+            models_of_kind
+            | 'Generate (key, count) for %ss' % self._kind >> beam.Map(
+                lambda model: ((self._kind, job_utils.get_model_id(model)), 1))
+        )
+
+
+class GetErrorsByExpectedModelKey(beam.PTransform):
+    """Returns PCollection of (key, error) pairs for each referenced model."""
+
+    def __init__(self, kind):
+        """Initializes the PTransform.
+
+        Args:
+            kind: str. The kind of model this PTransform will receive.
+        """
+        super(GetErrorsByExpectedModelKey, self).__init__(
+            label='Generate (key, error)s for models that %s depends on' % kind)
+        self._kind = kind
+        self._prop_references = tuple(PROP_REFERENCES_BY_KIND[kind].items())
+
+    def expand(self, models_of_kind):
+        """Returns PCollection of (key, error) pairs for each referenced model.
+
+        Args:
+            models_of_kind: PCollection. The models to inspect for property
+                references.
+
+        Yields:
+            PCollection. The (key, ModelRelationshipError) pairs of all models
+            referenced by a specific property on the input models. The key is a
+            (kind: str, ID: str) pair.
+        """
+        for prop_name, referenced_kinds in self._prop_references:
+            yield (
+                models_of_kind
+                | 'Generate errors from %s.%s' % (self._kind, prop_name) >> (
+                    beam.FlatMap(
+                        self._generate_expected_model_errors, prop_name,
+                        referenced_kinds))
+            )
+
+    def _generate_expected_model_errors(
+            self, model, prop_name, referenced_kinds):
+        """Yields all model keys referenced by the given model's properties.
+
+        Args:
+            model: Model. An individual input model.
+            prop_name: str. The property that holds the ID of another model.
+            referenced_kinds: tuple(str). The kinds of models that are expected
+                to correspond to the ID.
+
+        Yields:
+            tuple(tuple(str, str), ModelRelationshipError). The key for a
+            referenced model (kind: str, ID: str), and the error that should be
+            reported if the key does not exist.
+        """
+        prop_value = job_utils.get_model_property(model, prop_name)
+        if prop_value is None:
+            return
+        model_id, referenced_id = (
+            job_utils.get_model_id(model),
+            # IDs need to be in bytes to compare equal to an actual model's ID.
```

Comment:
I feel like you should probably do this in ModelRelationshipError instead -- in fact you already do that there for the model_id, right?